firmware/lib/Axis/src/Axis.cpp

```diff
@@ -0,0 +1,264 @@
+    pinMode(ENCODER_PIN_A_X, INPUT);
+    pinMode(ENCODER_PIN_B_X, INPUT);
+    Encoder encoder_x = {ENCODER_PIN_A_X, 10, 0};
+
+    pinMode(LIMIT_SW_HOME_PIN_X, INPUT_PULLUP);
+    LimitSwitch ls_home = {LIMIT_SW_HOME_PIN_X, digitalRead(LIMIT_SW_HOME_PIN_X)};
+
+    pinMode(LIMIT_SW_FAR_PIN_X, INPUT_PULLUP);
+    LimitSwitch ls_far_from_home = {LIMIT_SW_FAR_PIN_X, digitalRead(LIMIT_SW_FAR_PIN_X)};
+    
+    pinMode(DIR_PIN_X, OUTPUT);
+    pinMode(STEP_PIN_X, OUTPUT);
+    Axis axis_x_temp = {DIR_PIN_X, STEP_PIN_X, ACCEL_X, VEL_MAX_X, VEL_MIN_X, 
+                        {0, 0, 0}, 0, encoder_x, ls_home, ls_far_from_home,
+                        TC1, 0, 1, TC3_IRQn, TC4_IRQn};
+    return axis_x_temp;
+}
+
+Axis axis_x = setup_axis_x();
+
+// isr to handle encoder of x axis
+void isr_encoder_x()
+{
+    static unsigned long last_interrupt_time = 0;
+    unsigned long interrupt_time = millis();
+    // If interrupts come faster than 200ms, assume it's a bounce and ignore
+    if (interrupt_time - last_interrupt_time > 10) {
+        if (axis_x.dir == positive) axis_x.encoder.current++;
+        else if (axis_x.dir == negative) axis_x.encoder.current--;
+        print("Encoder ISR count: ", axis_x.encoder.current);
+    }
+    last_interrupt_time = interrupt_time;
+}
+
+// isr to handle limit switch
+void isr_limit_switch_x()
+{   
+    static unsigned long last_interrupt_time = 0;
+    unsigned long interrupt_time = millis();
+    // If interrupts come faster than 10ms, assume it's a bounce and ignore
+    if (interrupt_time - last_interrupt_time > 10) {
+        NVIC_DisableIRQ(axis_x.isr_velocity);
+        NVIC_DisableIRQ(axis_x.isr_accel);
+        print("X Home Limit switch has been hit", 1);
+
+        if (axis_x.homing && digitalRead(LIMIT_SW_HOME_PIN_X) == depressed) {
+            axis_x.encoder.current = 0;
+            axis_x.homing = 0;
+            print("X axis has been homed - SUCCESS", 1);
+        }
+        else if (axis_x.homing && digitalRead(LIMIT_SW_HOME_PIN_X) == pressed) {
+            home_axis(&axis_x);
+        }
+    }
+    last_interrupt_time = interrupt_time;
+}
+
+// isr for operating x axis motor velocity
+void TC3_Handler(void)
+{   
+    // print("Desired count: ", axis_x.encoder.desired);
+    // print("Current count: ", axis_x.encoder.current);
+    TC_GetStatus(TC1, 0);
+    PIOC->PIO_ODSR ^= PIO_ODSR_P25;
+    reset_timer(TC1, 0, TC3_IRQn, axis_x.vel); 
+}
+
+// isr for operating x axis motor acceleration 
+void TC4_Handler(void)
+{   
+    // print("Desired count: ", axis_x.encoder.desired);
+    // print("Current count: ", axis_x.encoder.current);
+    TC_GetStatus(TC1, 1);
+    // first accelerate
+    if (axis_x.tragectory_segment == accelerate) {
+        if (((axis_x.encoder.current < axis_x.encoder.desired) != axis_x.dir) && axis_x.vel < axis_x.vel_max) {
+            axis_x.vel++;
+            return;
+        }
+        else {
+            axis_x.tragectory_segment++;
```

Comment:
Are you incrementing an enum...I would classify that as relatively dangerous.
Especially since you didn't define numerical values for the enum constants.
I would say it's better to just hardcode what the next value is

Reply:
Not incrementing enum here, that's why segment is defined as int